File: api/schemas/project_setup.py
"""Versioned Project Setup schemas."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

from api.schemas.projects import ProjectPayload


class SetupValidationIssue(BaseModel):
    level: Literal["error", "warning"]
    field: Optional[str] = None
    message: str


class WorkflowReadiness(BaseModel):
    has_boundary: bool = False
    can_run_siteiq: bool = False
    can_run_terrainiq: bool = False
    can_run_layoutiq: bool = False
    can_run_yieldiq: bool = False


class ProjectSetupValidateResponse(BaseModel):
    valid: bool
    issues: List[SetupValidationIssue]
    readiness: WorkflowReadiness
    modules_to_run: List[str] = Field(default_factory=list)


class ProjectSetupValidateRequest(BaseModel):
    project_data: Dict[str, Any] = Field(default_factory=dict)


class ProjectPartialUpdateRequest(BaseModel):
    """Partial update — only provided top-level keys are merged."""

    name: Optional[str] = None
    center: Optional[Dict[str, float]] = None
    site_boundary_geojson: Optional[Dict[str, Any]] = None
    restriction_polygons_geojson: Optional[Dict[str, Any]] = None
    buildable_area_geojson: Optional[Dict[str, Any]] = None
    land_use: Optional[str] = None
    mount_type: Optional[str] = None
    country: Optional[str] = None
    workflow: Optional[Dict[str, Any]] = None
    project_info: Optional[Dict[str, Any]] = None
    location: Optional[Dict[str, Any]] = None
    geometry: Optional[Dict[str, Any]] = None
    design_basis: Optional[Dict[str, Any]] = None
    assumptions: Optional[Dict[str, Any]] = None
    schema_version: Optional[int] = None


class VersionedProjectPayload(ProjectPayload):
    schema_version: int = Field(default=1)
    project_info: Dict[str, Any] = Field(default_factory=dict)
    location: Dict[str, Any] = Field(default_factory=dict)
    geometry: Dict[str, Any] = Field(default_factory=dict)
    design_basis: Dict[str, Any] = Field(default_factory=dict)
    assumptions: Dict[str, Any] = Field(default_factory=dict)

File: frontend/src/types/projectSetup.ts
import type * as GeoJSON from "geojson";
import type { BoundaryPoint, LandUse } from "./gate";
import type { LayoutElectricalConfig, RoadMode } from "./layoutConfig";

export const PROJECT_SETUP_SCHEMA_VERSION = 1;

export type InputMethod =
  | "map"
  | "kml"
  | "kmz"
  | "geojson"
  | "shapefile"
  | "paste"
  | "import";

export type UnitsSystem = "metric" | "imperial";

export interface SetupParcel {
  id: string;
  name: string;
  layer_group: string;
  area_ha: number;
  coords: BoundaryPoint[];
  enabled: boolean;
}

export interface ProjectInfo {
  name: string;
  client: string;
  notes: string;
}

export interface ProjectLocation {
  country: string;
  state: string;
  city: string;
  lat: number;
  lon: number;
}

export interface ProjectGeometry {
  site_boundary?: BoundaryPoint[];
  parcels: SetupParcel[];
  restrictions: BoundaryPoint[][];
  buildable_area_geojson: GeoJSON.GeoJSON | null;
  buildable_area_ha: number | null;
  gross_area_ha: number;
}

export interface DesignBasis {
  land_use: LandUse;
  mount_type: string;
  target_capacity_mwp: number | null;
  target_cod: string;
  currency: string;
  coordinate_system: string;
  engineering_standard: string;
  design_standard: string;
  units: UnitsSystem;
}

export interface WorkflowReadiness {
  has_boundary: boolean;
  can_run_siteiq: boolean;
  can_run_terrainiq: boolean;
  can_run_layoutiq: boolean;
  can_run_yieldiq: boolean;
}

export interface ProjectSetupDraft {
  schema_version: number;
  project_info: ProjectInfo;
  location: ProjectLocation;
  geometry: ProjectGeometry;
  design_basis: DesignBasis;
  assumptions: LayoutElectricalConfig & {
    road_mode?: RoadMode;
    road_preset?: string;
    rows_per_block?: number;
    block_gap_m?: number;
  };
  input_method: InputMethod;
  workflow_state: {
    readiness: WorkflowReadiness;
  };
}

export interface SetupValidationIssue {
  level: "error" | "warning";
  field?: string;
  message: string;
}

export interface SetupValidationResult {
  valid: boolean;
  issues: SetupValidationIssue[];
  readiness: WorkflowReadiness;
  modules_to_run: string[];
}

export interface ProjectSetupValidateResponse {
  valid: boolean;
  issues: SetupValidationIssue[];
  readiness: WorkflowReadiness;
  modules_to_run: string[];
}
